agent/hcp/telemetry/filter.go

```diff
@@ -0,0 +1,42 @@
+package telemetry
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+// FilterList holds a map of filters, i.e. regular expressions.
+// These filters are used to identify which Consul metrics can be transmitted to HCP.
+type FilterList struct {
+	filters map[string]*regexp.Regexp
+}
+
+// NewFilterList returns a FilterList which holds valid regex
+// used to filter metrics. It will not fail if invalid REGEX is given, but returns a list of errors.
+func NewFilterList(filters []string) (*FilterList, error) {
+	var mErr error
+	compiledList := make(map[string]*regexp.Regexp, len(filters))
+	for idx, filter := range filters {
+		re, err := regexp.Compile(filter)
+		if err != nil {
+			mErr = multierror.Append(mErr, fmt.Errorf("compilation of filter at index %d failed: %w", idx, err))
+		}
+		compiledList[filter] = re
+	}
+	f := &FilterList{
+		filters: compiledList,
+	}
+	return f, mErr
+}
+
+// Match returns true if the metric name matches a REGEX in the allowed metric filters.
+func (fl *FilterList) Match(name string) bool {
+	for _, re := range fl.filters {
+		if re.Match([]byte(name)) {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
I'm a bit concerned about the performance of this. It's probably fine, since it's once every 15s. With this, though, we're doing a triple for loop. For each metric, for each pattern, does each character in the search string `name`.

This sounds kinda hair-brained but I wonder if it wouldn't be better to build up a single regex that's an OR of all the others like: `(p1)|(p2)|(p3)`. I think we'll be down to a double for loop because the regex's state machine will walk the metric just once for each search pattern (`name`)

And/or we could store cache on `FilterList` that stores whether we matched a given metric `name` to avoid using regex after the first match

Reply:
Good point, I was assuming we'd optimize the regex on the gateway side, but that makes total sense! I'll try to play with something and update that to a composed regex